### DISCUSSION TITLE
Training for Danish Support - Some experiences

### DISCUSSION THREAD

**Discussion by RJuro** [QwenLM/Qwen3-TTS#79] [General]
# Qwen3‑TTS Danish adaptation notes

**TL;DR**  
I’m trying to add **Danish** to Qwen3‑TTS using a single‑speaker CoRal‑TTS subset. I can clone the speaker’s **voice/timbre**, but Danish pronunciation is still **not intelligible**. The model works well in the **supported languages** with the voice trained in. I’m now testing **language adaptation** by unfreezing `talker.text_projection` + early talker layers with a low LR; loss drops, but output is still unstable/gibberish at times.

**Overall approach**  
- Start with the official pipeline (`sft_12hz.py`) to validate data + speaker cloning.  
- Then switch to a compact notebook for **selective unfreezing** (text→speech mapping layers) while keeping most of the backbone frozen.  
- Evaluate primarily by listening (short test sentences) after each epoch.

**What worked (training in a new voice ~voice cloning but only for supported languages)**  
- `sft_12hz.py` + local model path  
- `batch_size=2`, **LR around 2e‑6**, 3–10 epochs  

**What I’m trying now (not stable yet)**  
- Selective unfreezing: `talker.text_projection` + layers 0–1 (sometimes 0–3)  
- Lower LR (1e‑6–2e‑6), more epochs  
- A100 40GB, ~5k samples, short test prompts after each epoch

**Issues / fixes to consider**  
1) **Default LR too high** (`2e‑5` with batch=2 → noise). Recommend documenting lower LR.  
2) **Accelerate logging**: needs `project_dir` when `log_with="tensorboard"`.  
3) **HF model ID + `copytree`**: fails unless model is local.  
4) **Speaker name casing**: normalize to lowercase or document it.
5) **Possible official bug**: `sft_12hz.py` appears to **skip `talker.text_projection`** when building `inputs_embeds`. If `inputs_embeds` expects projected text features, this could hurt adaptation; worth clarifying in docs or fixing [`upstream.```]([url](url))

**Comment by MrMuhannadObeidat** [comment]
@RJuro , thanks for reporting the issues above. I ran into #3 and #2 in your issues list. I also tried to train for a new language (AR) with 10K samples with 2 different users. Adding to issues you mentioned, I am adding below:
1. What is expected to be passed for reference audio? It is not ideal to expect a reference voice for each unique speaker as once we go to bigger data sets such as mozilla common voice, list of speakers is unknown and widely varied.
2. Batching failed for me completely, did not diagnose, just brought down batch size to 1 as it was complaining about mismatch in tensor sizes.I would not be surprised if padding is required but not ideal.
3. Even though the model appears to have learned from 10K examples, it really does not speak intelligible Arabic.
4. I trained other models in the past like orpheus, sparktss and vibevoice and had much better success for proof of concepts runs immediately as compared to QWENTTS.

**Reply by RJuro** [reply]
nope, I gave up - looks like other parts of the model need to be trained for this to work.

**Reply by tekinek** [reply]
@MrMuhannadObeidat  
> I trained other models in the past like orpheus, sparktss and vibevoice and had much better success for proof of concepts runs immediately as compared to QWENTTS.

Could you please simply share your experience with these models?  (and does vibevoice support finetuning/new language adoptation?)

**Comment by torsteinelv** [comment]
Made sny progress on this? I have also tried to train in norwegian language but ending up with german accent wuth the the documwntet trainig script that looks to be only fintunkng the voice but npt the speach. Have not have any luck with this yet and currently im trying to fi tune the base modell to base model with lora hoping it will solve it.

**Reply by RJuro** [reply]
nope, gave up. Looks like it requires other parts of the models to be trained in other ways...

**Reply by torsteinelv** [reply]
ok. I managed to train the base model to speak norwegian, my training is here, but have not figured out the cloning part yet. 
https://github.com/torsteinelv/Qwen3-TTS-Norwegian/blob/main/src/train_norwegian_scratch_stable.py
but not happy with the quality of the result yet, so still doing some testing,

**Comment by MrMuhannadObeidat** [comment]
@torsteinelv, thanks for providing your training setup. I cloned that and tried to use it against an Arabic data set. My dataset right now only contains 10K samples and I ran for 5 epochs. No issues to get the script to work and produce the different checkpoints. Loss came down to around 2.3 from 13 starting point. However, the resulting model produces gibberish. 
I used the original qwen script to create the jsonl file and used following parameter in your script for training (all defaults):
export NUM_EPOCHS=5
export BATCH_SIZE=4
export GRAD_ACCUM=4

export LORA_LR="1e-5"
export TEXT_PROJ_LR="5e-5"
export SUB_LOSS_WEIGHT="0.2"

export LORA_R=4
export LORA_ALPHA=-8
export LORA_DROPOUT=-0.1
export TRAIN_MLP_LORA=false

export MIXED_PRECISION=bf16

Qwen/Qwen3-TTS-12Hz-1.7B-Base

**Reply by torsteinelv** [reply]
did you use: https://github.com/torsteinelv/Qwen3-TTS-Norwegian/blob/main/src/test_lora.py when testing it? I do get norwegian accent but having problems with it talking to fast now. 

Im close to give this up as its to difficult to figure out and cross fingers that qwen can share a way to do this later. btw I did train on 9 hours first, then I some time could hear norwegian words clear with alot of noice when it happend. last try now is about 150 hours of data and its clear but having some speed issues.

**Reply by Tomiinek** [reply]
Hi @torsteinelv I think I am experiencing the same struggle. I would like to share my findings with you since I am also very close to giving up. 

I experimented with fine-tuning on Czech, >1k hours of data and I am quite sure it has good quality. 
- when tuning the model as a whole, I was getting gibberish on output (despite large batch sizes, small and high learning rates or whatever ...)
- so I started with tuning of the text embedding and projection, it went fine, accent was diminished, but still very harsh
- so I added also gentle backbone fine-tuning (with lora or without, it does not matter), but the output speech is getting faster and faster (not like sampling rate issue, but really like faster tempo with ok timbre) and also the EOS is starting to be ignored.  
- so I decided to make a step back, and tuned the model on LJ Speech (full fine-tuning, normal hparams). It just works fine.
- so I decided to fine-tune the model on LibriTTS-R (multi speaker), but it does not work for me. I am using the same setup as for LJ Speech, using a different speaker embedding for each batch item ... 🤷 The training cruves are kind of the same as for LJ.

<img width="1517" height="316" alt="image" src="https://github.com/user-attachments/assets/e5114d17-385d-4997-9448-23e845270482" />

**Reply by RJuro** [reply]
Very cool, thanks for the update. Had same results - also the speedup that you mention.

**Reply by torsteinelv** [reply]
For me I feel like the earlier epochs have the best results in some way also. and I have got rid of gibberish output in the latest try I had but now only speeding was able to get rid of this aslo with experimenting with repetition_penalty but its not perfekt. In the begging in the output it starts fine, and as it comes closed to the end it speeds more up if the clip has long output.

**Reply by torsteinelv** [reply]
@Tomiinek  are you traing to voiceclone or base to base model?

**Reply by MrMuhannadObeidat** [reply]
@torsteinelv Thanks for all the guidance, made it much better progress now.

I was not using the test script, I now converted into a notebook to do a bit of interactive testing and now getting Arabic voice correctly. I am now using your training script and test script equivalent, starting from base model:
Qwen3-TTS-12Hz-1.7B-Base

Voice quality is good after 5 epochs on a 10K audio sample data set. The average length of audio is 4seconds. I have the same issue with the speed of speech that all others referred to here. If anyone has tips on how to address then great.

Next I will do a training run on bigger corups and have to look into two areas:
1. Merge lora adatpers so I can do more than one training phases with different datasets of varying qualities.
2. Deal with issues in training script related to running stack overflow when a data set is big. My generic data set is 140K samples and the script throws a recursion exception.

**Reply by Tomiinek** [reply]
> @Tomiinek are you traing to voiceclone or base to base model?

I believe I am doing base-to-base. I use `x_vector_only_mode=False` and no ICL. Have you tried some kind of ICL?

> I had but now only speeding was able to get rid of this aslo with experimenting with repetition_penalty but its not perfekt. In the begging in the output it starts fine, and as it comes closed to the end it speeds more up if the clip has long output.

Yeah, even the LJ Speech tuning was not effortless. I had to tune the inference parameters a bit. The defaults resulted in a terrible speech and ofc omitting EOS. What seemed to be quite stable is disabled sampling on talker and lower temperature and 0.95 topp on subtalker. It also probably speeds a bit up at the end. Long text inputs just do not work (but it makes sense given the training data).

**Reply by torsteinelv** [reply]
Cool please update if you figure something out :) I

**Reply by Tomiinek** [reply]
FYI https://github.com/QwenLM/Qwen3-TTS/pull/178

**Reply by torsteinelv** [reply]
Have you tried? I will try to correct this and run a new training and see if this solves it. 🙂 id it works we can try to merge the models after.

**Reply by Tomiinek** [reply]
Here we go: https://github.com/QwenLM/Qwen3-TTS/issues/179#issuecomment-3860084310

**Reply by torsteinelv** [reply]
So with this fix its not a ploblem training a new language?

**Comment by arpit763** [comment]
can anyone help me with merging the adapter with base model so that u can further finetune it?

currently i am using this script which only allows me to do inference not stage 2 finetunung.

`import json
import logging
import shutil
from pathlib import Path

import torch
from peft import PeftModel
from peft.tuners.lora import LoraLayer
from huggingface_hub import save_torch_model

from qwen_tts import Qwen3TTSModel


# ==============================================================================
# CONFIG
# ==============================================================================

BASE_MODEL_PATH = "/artifacts/Qwen3-TTS-12Hz-1.7B-Base"
ADAPTER_PATH = "/artifacts/shona_stage1/checkpoint-epoch-9"
MERGED_OUTPUT_DIR = "/artifacts/qwen3-tts-1.7b-merged-scale02"

LORA_SCALE = 0.2

DTYPE = torch.bfloat16
ATTN_IMPLEMENTATION = "flash_attention_2"

# ==============================================================================
# LOGGING
# ==============================================================================

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s | %(levelname)-8s | %(message)s",
)

logger = logging.getLogger(__name__)


# ==============================================================================
# HELPERS
# ==============================================================================

def copy_file(src: Path, dst: Path):
    if src.exists():
        shutil.copy2(src, dst)
        logger.info(f"Copied {src.name}")
    else:
        logger.warning(f"Missing {src}")


def set_lora_scale(peft_model, scale: float):

    logger.info(f"Applying LoRA scale = {scale}")

    if scale == 1.0:
        logger.info("Scale is 1.0, skipping.")
        return

    adapter = getattr(peft_model, "active_adapter", None)

    count = 0

    for module in peft_model.modules():

        if isinstance(module, LoraLayer):

            if adapter is None:
                module.scale_layer(scale)

            elif isinstance(adapter, (list, tuple)):
                for name in adapter:
                    module.set_scale(name, scale)

            else:
                module.set_scale(adapter, scale)

            count += 1

    logger.info(f"Applied scale to {count} LoRA layers.")


# ==============================================================================
# MAIN
# ==============================================================================

def main():

    output_dir = Path(MERGED_OUTPUT_DIR)
    output_dir.mkdir(parents=True, exist_ok=True)

    logger.info("=" * 80)
    logger.info("Loading base model...")
    logger.info("=" * 80)

    tts = Qwen3TTSModel.from_pretrained(
        BASE_MODEL_PATH,
        torch_dtype=DTYPE,
        device_map="cpu",
        attn_implementation=ATTN_IMPLEMENTATION,
    )

    logger.info("Base model loaded.")

    logger.info("=" * 80)
    logger.info("Loading LoRA adapter...")
    logger.info("=" * 80)

    peft_model = PeftModel.from_pretrained(
        tts.model,
        ADAPTER_PATH,
    )

    logger.info("Adapter loaded.")

    # --------------------------------------------------------------------------
    # APPLY LORA SCALE
    # --------------------------------------------------------------------------

    set_lora_scale(peft_model, LORA_SCALE)

    logger.info("=" * 80)
    logger.info("Merging adapter...")
    logger.info("=" * 80)

    merged_model = peft_model.merge_and_unload()

    merged_model.eval()

    logger.info("Merge complete.")

    logger.info("=" * 80)
    logger.info("Saving merged weights...")
    logger.info("=" * 80)

    save_torch_model(
        merged_model,
        str(output_dir),
        safe_serialization=True,
    )

    logger.info("Weights saved.")

    # --------------------------------------------------------------------------
    # Processor / tokenizer
    # --------------------------------------------------------------------------

    if hasattr(tts, "processor"):
        logger.info("Saving processor...")
        tts.processor.save_pretrained(output_dir)

    # --------------------------------------------------------------------------
    # Generation config
    # --------------------------------------------------------------------------

    copy_file(
        Path(BASE_MODEL_PATH) / "generation_config.json",
        output_dir / "generation_config.json",
    )

    speech_src = Path(BASE_MODEL_PATH) / "speech_tokenizer"
    speech_dst = output_dir / "speech_tokenizer"

    if speech_src.exists():
        if speech_dst.exists():
            shutil.rmtree(speech_dst)

        shutil.copytree(speech_src, speech_dst)
        logger.info("Copied speech_tokenizer/")
    else:
        logger.warning("speech_tokenizer not found!")

    # --------------------------------------------------------------------------
    # Adapter config
    # --------------------------------------------------------------------------

    adapter_cfg = Path(ADAPTER_PATH) / "config.json"

    if adapter_cfg.exists():

        with open(adapter_cfg, "r", encoding="utf-8") as f:
            cfg = json.load(f)

        with open(output_dir / "config.json", "w", encoding="utf-8") as f:
            json.dump(
                cfg,
                f,
                indent=2,
                ensure_ascii=False,
            )

        logger.info("Saved config.json")

    else:
        logger.warning("Adapter config.json not found.")

    # --------------------------------------------------------------------------
    # Speaker embedding
    # --------------------------------------------------------------------------

    copy_file(
        Path(ADAPTER_PATH) / "speaker_embedding.safetensors",
        output_dir / "speaker_embedding.safetensors",
    )

    logger.info("=" * 80)
    logger.info("DONE")
    logger.info(f"Merged model saved to:\n{output_dir}")
    logger.info("=" * 80)


if __name__ == "__main__":
    main()


`
